src/compression/mod.rs

```diff
@@ -0,0 +1,176 @@
+    })
+}
+
+/// struct that knows how to decompress LAZ
+///
+/// Decompression is done in 2 step:
+/// 1) call the decompressor the read & decompress the next point
+/// and put its data in a in-memory buffer
+/// 2) read the buffer to get the decompress point
+pub(crate) struct CompressedPointReader<R: Read +Seek> {
+    /// decompressor that does the actual job
+    decompressor: laz::las::laszip::LasZipDecompressor<R>,
+    header: Header,
+    /// in-memory buffer where the decompressor writes deompression result
+    decomp_out: Cursor<Vec<u8>>,
+    last_point_idx: u64,
+}
+
+impl<R: Read + Seek> CompressedPointReader<R> {
+    pub(crate) fn new(source: R, mut header: Header) -> Result<Self> {
+        let mut index = None;
+        for (i, vlr) in header.vlrs().iter().enumerate() {
+            if &vlr.user_id == LASZIP_USER_ID && vlr.record_id == LASZIP_RECORD_ID {
+                index = Some(i);
+            }
+        }
+        let index = index.expect("Data is compressed but no Laszip VLR found");
+        let laszip_vlr = header.vlrs_mut().remove(index);
+        let laszip_vlr = laz::las::laszip::LazVlr::from_buffer(&laszip_vlr.data)?;
+        let decomp_out = Cursor::new(vec![0u8; header.point_format().len() as usize]);
+
+        Ok(Self {
+            decompressor: laz::las::laszip::LasZipDecompressor::new(source, laszip_vlr)?,
+            header,
+            decomp_out,
+            last_point_idx: 0,
+        })
+    }
+}
+
+impl<R: Read + Seek> Debug for CompressedPointReader<R> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "CompressedPointReader(num_read: {}, header: {:?})", self.last_point_idx, self.header)
+    }
+}
+
+impl<R: Read + Seek> PointReader for CompressedPointReader<R> {
+    fn read_next(&mut self) -> Option<Result<Point>> {
+        if self.last_point_idx < self.header.number_of_points() {
+            self.last_point_idx += 1;
+            self.decompressor.decompress_one(&mut self.decomp_out.get_mut()).unwrap();
+            if let Err(e) = self.decomp_out.seek(SeekFrom::Start(0)) {
+                Some(Err(e.into()))
+            } else {
+                Some(read_point_from(&mut self.decomp_out, &self.header))
+            }
+        } else {
+
+            None
+        }
+    }
+
+    fn seek(&mut self, position: u64) -> Result<()> {
+        self.last_point_idx = position - 1;
+        self.decompressor.seek(position)?;
+        Ok(())
+    }
+
+    fn header(&self) -> &Header {
+        &self.header
+    }
+}
+
+
+/// struct that knows how to write LAZ
+///
+/// Writing a point compressed is done in 2 steps
+/// 1) write the point to a in-memory buffer
+/// 2) call the laz compressor on this buffer
+pub(crate) struct CompressedPointWriter<W: Write + Seek> {
+    header: Header,
+    /// buffer used to write the un compressed point
+    compression_in: Cursor<Vec<u8>>,
+    /// The compressor that actually doest the job of compressing the data
```

Comment:
Punctuation.

Reply:
Appart from the typo on `doest`, I don't see where the punctuation is missing  😧

Reply:
Ah yup you're right, typo not punctuation, my bad.